### PR TITLE
Grammatical Fixes in qaoa.ipynb

### PR DESCRIPTION
**Pull request by davisclarke**
1. Added proper punctuation for the opening bullet points (bullet point punctuation with longer bullet points.
2. QAOA (Quantum Approximate Optimization Algorithm) introduced by Farhi et al .[1] -> QAOA (Quantum Approximate Optimization Algorithm) was introduced by Farhi et al [1].
3. For that we would first define
the underlying graph of the problem shown above. -> For that, we would first define the underlying graph of the problem as shown above.
4. Such a state, when number of qubits is 4 ($n=4$), can be prepared by applying Hadamard gates starting from an all zero state as shown in 
the circuit below. -> Such a state, when the number of qubits is 4 ($n=4$), can be prepared by applying Hadamard gates starting from an all-zero state as shown in the circuit below.
5. So far we have seen that the preparation of a quantum state during QAOA is composed of three elements -> So far, we have seen that the preparation of a quantum state during QAOA is composed of three elements:
6. Such an expectation can be obtained by doing measurement in the Z-basis. -> Such an expectation can be obtained by doing measurements in the Z-basis.
7. Hence the general form a combinatorial optimization problem is given by -> Hence, the general form a combinatorial optimization problem is given by
8. This cost function can be mapped to a Hamiltonian that is diagonal in the computational basis. Given the cost-function $C$ this Hamiltonian is then written as -> This cost function can be mapped to a Hamiltonian that is diagonal in the computational basis. Given the cost function, $C$, this Hamiltonian is then written as
9. The cost function to be optimized is in this case the sum of weights of edges connecting points in the two different subsets, _*crossing*_ the cut. -> The cost function to be optimized is, in this case, the sum of weights of edges connecting points in the two different subsets, _*crossing*_ the cut.
10.  This can of course be turned again into a decision problem if we ask where there exists a bit string that satisfies more than $\tilde{m}$ of the $m$ clauses, which is again $NP$-complete. -> Of course, this can be turned again in to a decision problem if we ask where there exists a bit string that satisfies more than $\tilde{m}$ of the $m$ clauses, which is again $NP$-complete. (Redundancy)
11. Both the previously considered problems $MAXCUT$ and $\text{MAX 3-SAT}$ are actually known to be a NP-hard problems [3](#references). -> Both the previously considered problems $MAXCUT$ and $\text{MAX 3-SAT}$ are known to be a NP-hard problems [3](#references). (Redundancy)
12. In fact it turns out that many combinatorial optimization problems are computationally hard to solve in general -> In fact, it turns out that many combinatorial optimization problems are computationally hard to solve in general
13. In light of this fact, we can't expect to find a provably efficient algorithm, i.e. an algorithm with polynomial runtime in the problem size, that solves these problems. This also applies to quantum algorithms. There are two main approaches to dealing with such problems. First approach is approximation algorithms that are guaranteed to find solution of specified quality in polynomial time. The second approach are heuristic algorithms that don't have a polynomial runtime guarantee but appear to perform well on some instances of such problems. -> In light of this, we can't expect to find a provably efficient algorithm, i.e., an algorithm with polynomial runtime in the problem size, that solves these problems; this also applies to quantum algorithms. There are two main approaches to dealing with such problems. First approach is approximation algorithms that are guaranteed to find solution of specified quality in polynomial time. The second approach, heuristic algorithms, don't have a polynomial runtime guarantee, but appear to perform well on some instances of such problems. (Clarity, redundancy, comma rules)
14. For the $MAXCUT$ problem there is a famous approximate algorithm due to Goemans and Williamson [2](#references) .-> For the $MAXCUT$ problem, there is a famous approximate algorithm due to Goemans and Williamson [2](#references).
15. This algorithm is based on an SDP relaxation of the original problem combined with a probabilistic rounding technique that yields an with high probability approximate solution $\textbf{x}^*$ that has an approximation ratio of $\alpha \approx 0.878$. This approximation ratio is actually believed to optimal so we do not expect to see an improvement by using a quantum algorithm.-> This algorithm, based on an SDP relaxation of the original problem, is combined with a probabilistic rounding technique that yields a high probability approximate solution $\textbf{x}^*$ that has an approximation ratio of $\alpha \approx 0.878$. This approximation ratio is believed to be optimal. So, we do not expect to see an improvement by using a quantum algorithm.
16. The Quantum approximate optimization algorithm (QAOA) -> The Quantum Approximate Optimization Algorithm (QAOA)
17. QAOA takes the approach of classical approximate algorithms and looks for a quantum analogue that will likewise produce a classical bit string $x^*$ that with high probability is expected to have a good approximation ratio $\alpha$. -> QAOA takes the approach of classical approximate algorithms and looks for a quantum analogue that will likewise produce a classical bit string $x^*$. The outputted bit string is expected, with a high probability, to have a good approximation ratio $\alpha$. (Run-on sentence)

### CODE REVIEW

**Comment by davisclarke** [comment]
Sorry. For some reason, an [already merged commit](https://github.com/Qiskit/textbook/pull/45) showed up here.

File: notebooks/ch-applications/qaoa.ipynb [hunk 1]
```diff
@@ -53,12 +58,13 @@
...
   },
   {
+   "attachments": {},
    "cell_type": "markdown",
    "metadata": {},
    "source": [
     "## QAOA\n",
     "\n",
-    "QAOA (Quantum Approximate Optimization Algorithm) introduced by Farhi et al.[1] is a quantum algorithm that attempts to \n",
+    "QAOA (Quantum Approximate Optimization Algorithm) was introduced by Farhi et al [1]. is a quantum algorithm that attempts to \n",
```
Comment (frankharkins) [inline notebooks/ch-applications/qaoa.ipynb:67]:
```suggestion
    "QAOA (Quantum Approximate Optimization Algorithm), introduced by Farhi et al. [1], is a quantum algorithm that attempts to \n",
```

Couple of things here:
 - "Et al." needs the period as it's an abbreviation
 - The "introduced by..." is an interjection, so I've removed the "was" and surrounded with commas.

File: notebooks/ch-applications/qaoa.ipynb [hunk 2]
```diff
@@ -667,18 +687,19 @@
     "\n",
     "\n",
     "\n",
-    "For the $MAXCUT$ problem there is a famous approximate algorithm due to Goemans and Williamson [2](#references) . This algorithm is based on an SDP relaxation of the original problem combined with a probabilistic rounding technique that yields an with high probability approximate solution $\\textbf{x}^*$ that has an approximation ratio of $\\alpha \\approx 0.878$. This approximation ratio is actually believed to optimal so we do not expect to see an improvement by using a quantum algorithm."
+    "For the $MAXCUT$ problem, there is a famous approximate algorithm due to Goemans and Williamson [2](#references).This algorithm, based on an SDP relaxation of the original problem, is combined with a probabilistic rounding technique that yields a high probability approximate solution $\\textbf{x}^*$ that has an approximation ratio of $\\alpha \\approx 0.878$. This approximation ratio is believed to be optimal. So, we do not expect to see an improvement by using a quantum algorithm."
```
Comment (frankharkins) [inline notebooks/ch-applications/qaoa.ipynb:690]:
```suggestion
    "For the $MAXCUT$ problem, there is a famous approximate algorithm due to Goemans and Williamson [2](#references). This algorithm, based on an SDP relaxation of the original problem, is combined with a probabilistic rounding technique that yields a high probability approximate solution $\\textbf{x}^*$ that has an approximation ratio of $\\alpha \\approx 0.878$. This approximation ratio is believed to be optimal. So, we do not expect to see an improvement by using a quantum algorithm."
```

File: notebooks/quantum-machine-learning/qgan.ipynb [hunk 3]
```diff
@@ -279,7 +283,7 @@
...
     "\n",
-    "The objective with a fully quantum QGAN is for the quantum generator to reproduce a desired state $|\\psi\\rangle$, using of course an adversarial training strategy. Similar to other variational quantum algorithms, the quantum generator moves towards this through an iterative update to its parameters directed by a classical optimizer. However, in the case of a QGAN, the generator's cost function landscape itself evolves and becomes better as the discriminator improves at recognizing real samples. Let's look at the general circuit schematics of what we will be building.\n",
+    "The objective of a fully quantum QGAN is for the quantum generator to reproduce a desired state $|\\psi\\rangle$, of course using an adversarial training strategy. Similar to other variational quantum algorithms, the quantum generator moves towards this through an iterative update to its parameters directed by a classical optimizer. However, in the case of a QGAN, the generator's cost function landscape evolves by itself and becomes better as the discriminator improves at recognizing real samples. Let's look at the general circuit schematics of what we will be building.\n",
```
Comment (frankharkins) [inline notebooks/quantum-machine-learning/qgan.ipynb]:
```suggestion
    "The objective of a fully quantum QGAN is for the quantum generator to reproduce a desired state $|\\psi\\rangle$, of course using an adversarial training strategy. Similar to other variational quantum algorithms, the quantum generator moves towards this through an iterative update to its parameters directed by a classical optimizer. However, in the case of a QGAN, the generator's cost function landscape also evolves and becomes better as the discriminator improves at recognizing real samples. Let's look at the general circuit schematics of what we will be building.\n",
```

This changed the meaning a bit; the original sentence is saying that the generator cost function evolves, whereas in the past we've assumed the generator cost functions are fixed. Let me know if you think my suggested change clears that up.

**Review by frankharkins — CHANGES_REQUESTED** [review]
Thanks! Most changes are great, I've just left a couple of comments.

File: notebooks/ch-applications/qaoa.ipynb (hunk 1 above)
Comment (davisclarke) [reply notebooks/ch-applications/qaoa.ipynb:67]: My mistake! I was trying to it up into two sentences; I should have reviewed this. My apologies, having the "...introduced by" as an interjection reads better.

File: notebooks/ch-applications/qaoa.ipynb (hunk 2 above)
Comment (davisclarke) [reply notebooks/ch-applications/qaoa.ipynb:690]: Once again, simple error on my part.

File: notebooks/quantum-machine-learning/qgan.ipynb (hunk 3 above)
Comment (davisclarke) [reply notebooks/quantum-machine-learning/qgan.ipynb]: I see. Resolved in last commit.

File: notebooks/ch-applications/qaoa.ipynb (hunk 1 above)
Comment (frankharkins) [reply notebooks/ch-applications/qaoa.ipynb]: No worries! Thanks

**Review by frankharkins — APPROVED** [review]
Looks good, thanks!

**Pull request merged**